pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	kmetrics "k8s.io/component-base/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+func init() { //nolint:gochecknoinits // metrics should be registered once
+	metrics.Registry.MustRegister(drift, mr, mrReady, mrSynced, mrDetected, mrReadyDuration, mrDeletionDuration)
+}
+
+const subSystem = "crossplane"
+
+var (
+	drift = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "resource_drift_seconds",
+		Help:      "ALPHA: How long since the previous successful reconcile when a resource was found to be out of sync; excludes restart of the provider",
+		Buckets:   kmetrics.ExponentialBuckets(10e-9, 10, 10),
+	}, []string{"group", "kind"})
+
+	mr = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_created",
+		Help:      "The number of managed resources created",
+	}, []string{"gvk", "name", "claim", "composite"})
```

Comment:
Yes, that was my first attempt :) The problem I faced again was sending multiple metrics for the same MR each time the condition combination updated.

Reply:
This could be a sign that this reconciler isn't the right place to be tracking these metrics.

Kubernetes frames this well [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/metric-instrumentation.md#instrumentation-types):

> Components have metrics capturing events and states that are inherent to their application logic. Examples are request and error counters, request latency histograms, or internal garbage collection cycles. Those metrics are instrumented directly in the application code.
>
> Secondly, there are business logic metrics. Those are not about observed application behavior but abstract system state, such as desired replicas for a deployment. They are not directly instrumented but collected from otherwise exposed data.

In the context of this PR, we could ask:

* Which metrics are helping us understand how well the MR reconciler you're instrumenting is behaving?
* Which metrics are instead telling us about the state of the system as a whole?

I would argue these metrics are telling us how the reconciler is behaving:

* managed_resource_detected_time_seconds - Average time to notice a resource exists
* managed_resource_ready_duration_seconds - Average time to drive a resource to readiness
* managed_resource_deletion_seconds - Average time to delete a resource
* resource_drift_seconds - Average time to correct drift for a resource

Whereas these metrics are just telling us about the state of the system:

* managed_resource_created - Number of resources that exist
* managed_resource_ready - Number of resources that are ready
* managed_resource_synced - Number of resources that are synced

In Kubernetes, kube-state-metrics would monitor things like this:

> In Kubernetes they are generally captured in the [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) component, which reads them from the API server. For this types of metric exposition, the [exporter guidelines](https://prometheus.io/docs/instrumenting/writing_exporters/) apply additionally.

There's prior art for kube-state-metrics for Crossplane in https://github.com/crossplane-contrib/x-metrics.

I'm not necessarily making the case that these metrics should move into x-metrics. It would be more convenient if you didn't need to run an extra component to get access to things like resource counts, especially given our resources have a lot of dynamic types.

One way to do handle this could be to add a new crossplane-runtime package that a provider can import to expose KSM-style metrics. I could imagine rather than being a controller, such a package would simply occasionally list a particular kind of MR and expose metrics about how many existed, were ready, etc.

Reply:
Thanks for your detailed review and sharing all these links. They helped a lot! It explains why I was struggling to record those in the reconciler. 

I've moved the state metrics from the reconciler and recording them in a goroutine in the `NewReconciler` function now. But I'm not sure if there might be a better place to call it.